pkg/reconciler/v1alpha1/pipelinerun/resources/pipelinestate.go

```diff
@@ -54,8 +54,32 @@ func GetNextTask(prName string, state []*PipelineRunTaskRun, logger *zap.Sugared
 	return nil
 }
 
-func canTaskRun(pt *v1alpha1.PipelineTask) bool {
+func canTaskRun(pt *v1alpha1.PipelineTask, state []*PipelineRunTaskRun) bool {
 	// Check if Task can run now. Go through all the input constraints
+	for _, input := range pt.InputSourceBindings {
+		if len(input.PassedConstraints) > 0 {
+			for _, constrainingTaskName := range input.PassedConstraints {
+				for _, prtr := range state {
+					// the constraining task must have a successful task run to allow this task to run
+					if prtr.Task.Name == constrainingTaskName {
+						if prtr.TaskRun == nil {
+							return false
+						}
+						c := prtr.TaskRun.Status.GetCondition(duckv1alpha1.ConditionSucceeded)
+						if c == nil {
+							return false
+						}
+						switch c.Status {
+						case corev1.ConditionFalse:
+							return false
+						case corev1.ConditionUnknown:
+							return false
+						}
+					}
+				}
+			}
+		}
+	}
```

Comment:
whoa nice! im curious, did you need to add this for the helm deployment to work or was this just a bonus?

One thing I am a bit confused about: let's say we have these tasks:

1. Task A
2. Task B - has a passed constraint on task A

Since Task A has to run before Task B, how can this function ever return False? This might be more relevant once we are constructing a DAG instead of executing Tasks in the order they are declared in, but even then, I think the `passedConstraints` would be used to construct that DAG only.

Long story short, I think if this function returns `False`, it means the `PipelineRun` is in an error state and will never complete, and we need to return an error and stop reconciling the `PipelineRun` at that point (maybe I'm missing something tho!)

Reply:
I thought the test was failing because both tasks were running at the same time, and since the deploy one needs the image from the first one, I added this functionality 
my understanding was that the `cantaskrun` will return false if the task is not ready, but will get called again in the following reconcile period, am I wrong?

Reply:
umm. may be some misunderstanding.  The `GetNextTask` is still check if the status of the previous task run is complete.
Ideally, we would not care if the prev task is completed or not. 
```
func GetNextTask(prName string, state []*PipelineRunTaskRun, logger *zap.SugaredLogger) *PipelineRunTaskRun {
 	for _, prtr := range state {
 		if prtr.TaskRun != nil {
 			c := prtr.TaskRun.Status.GetCondition(duckv1alpha1.ConditionSucceeded)
 			if c == nil {
 				logger.Infof("TaskRun %s doesn't have a condition so it is just starting and we shouldn't start more for PipelineRun %s", prtr.TaskRunName, prName)
 				return nil
 			}
 			switch c.Status {
 			case corev1.ConditionFalse:
 				logger.Infof("TaskRun %s has failed; we don't need to run PipelineRun %s", prtr.TaskRunName, prName)
 				return nil
 			case corev1.ConditionUnknown:
				logger.Infof("TaskRun %s is still running so we shouldn't start more for PipelineRun %s", prtr.TaskRunName, prName)
				return nil
			}
		} else if canTaskRun(prtr.PipelineTask) {
		} else if canTaskRun(prtr.PipelineTask, state) {
			logger.Infof("TaskRun %s should be started for PipelineRun %s", prtr.TaskRunName, prName)
			return prtr
		}
}
```